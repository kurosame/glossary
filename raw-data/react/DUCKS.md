## category

react

## titles

Ducks
Redux

## description

Redux のデザインパターン  
Action や Reducer を modules にパッケージする

vuejs-boilerplate でも使ってる気がする  
1 ファイルにはしてないけど、actions, getters, mutations, state を import して modules でパッケージして export してる  
そして Store 作るときに modules をセットしている
