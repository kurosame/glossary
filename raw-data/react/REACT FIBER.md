## category

react

## titles

React Fiber
react-fiber

## description

Begin, Complete, Commit のフェーズがある

### Begin, Complete

Component 単位で実行される  
componentWillMount, componentWillReceiveProps, shouldComponentUpdate, componentWillUpdate  
非同期レンダリングでは中断・再開により、複数回ライフサイクル関数が呼ばれる可能性がある

### Commit

全ての Fiber が完了した時に実行される  
DOM 更新などの副作用が発生する処理はここでまとめて反映される  
componentDidMount, componentDidUpdate, componentWillUnmount

### Time Slicing

Fiber により画面のレンダリングが非同期になると、Fiber 単位で処理の優先度を付けることができる  
例えばユーザ操作の優先度を高くすることで、画面がレンダリング中でもユーザ操作を優先して処理することが可能となる

### Suspense

非同期通信をしている間、画面のレンダリングを中断できる  
非同期通信が完了した後に、画面をレンダリングすることが可能  
render に Promise を投げて、その Promise が完了した後に render するみたいな？
