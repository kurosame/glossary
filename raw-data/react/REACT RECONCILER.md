## category

react

## titles

React Reconciler
react-reconciler

## description

VDOM と呼ばれている部分

1. ReactElement のツリーを構築
1. Component をマウント、アンマウント、更新
1. ライフサイクル関数を呼出す
1. 現在の DOM との差分を計算する
1. renderer に差分を伝える
