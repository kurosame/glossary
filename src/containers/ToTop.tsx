import React from 'react'

import { KeyboardArrowUp } from '@mui/icons-material'
import { Box as MBox, Fab, useScrollTrigger, Zoom } from '@mui/material'
import { styled } from '@mui/system'

const Box = styled(MBox)`
  position: fixed;
  right: 32px;
  bottom: 32px;
`

const ToTop: React.FC = () => (
  <Zoom in={useScrollTrigger({ disableHysteresis: true })}>
    <Box>
      <Fab
        color="primary"
        size="small"
        data-testid="to-top"
        onClick={() => {
          window.scroll({ top: 0 })
        }}
      >
        <KeyboardArrowUp />
      </Fab>
    </Box>
  </Zoom>
)

export default ToTop
