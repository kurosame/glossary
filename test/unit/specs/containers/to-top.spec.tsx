/**
 * @jest-environment jsdom
 */
import React from 'react'

import { cleanup, fireEvent, render, RenderResult } from '@testing-library/react'

import ToTop from '@/containers/ToTop'

let wrapper: RenderResult
beforeEach(() => {
  wrapper = render(<ToTop />)
  window.scroll = jest.fn()
})
afterEach(cleanup)

test('Match the snapshot', () => {
  expect(wrapper.asFragment()).toMatchSnapshot()
})

test('Match the snapshot when click', () => {
  fireEvent.click(wrapper.getByTestId('to-top'))

  expect(wrapper.asFragment()).toMatchSnapshot()
})
