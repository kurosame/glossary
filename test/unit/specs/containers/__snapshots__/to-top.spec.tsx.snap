// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1jfyc7e"
    style="transform: scale(0); visibility: hidden;"
  >
    <button
      class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-primary css-1dm1o0z-MuiButtonBase-root-MuiFab-root"
      data-testid="to-top"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="KeyboardArrowUpIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41 15.41 12 10.83l4.59 4.58L18 14l-6-6-6 6z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`Match the snapshot when click 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1jfyc7e"
    style="transform: scale(0); visibility: hidden;"
  >
    <button
      class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-primary css-1dm1o0z-MuiButtonBase-root-MuiFab-root"
      data-testid="to-top"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="KeyboardArrowUpIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41 15.41 12 10.83l4.59 4.58L18 14l-6-6-6 6z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
